pkg/cli/image/strategy/simplelookup.go

```diff
@@ -0,0 +1,166 @@
+	return nil, nil
+}
+
+// OnFailure returns a list of possible image references for the locator image reference, gathered from image content source policies, or an error
+func (s *simpleLookupICSP) OnFailure(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if err := s.resolve(ctx, locator); err != nil {
+		return nil, err
+	}
+	if len(s.alternates) == 0 {
+		return nil, fmt.Errorf("no alternative image references found for image: %s", locator.String())
+	}
+	return s.alternates, nil
+}
+
+// resolve gathers possible image sources for a given image
+// gathered from ImageContentSourcePolicy objects and user-passed image.
+// Will lookup from cluster or from ImageContentSourcePolicy file passed from user.
+// Image reference of user-given image may be different from original in case of mirrored images.
+func (s *simpleLookupICSP) resolve(ctx context.Context, imageRef reference.DockerImageReference) error {
+	if s.alternates != nil {
+		return nil
+	}
+
+	var icspList []operatorv1alpha1.ImageContentSourcePolicy
+	var err error
+	if len(s.icspFile) > 0 {
+		klog.V(5).Infof("Reading ICSP from file %s", s.icspFile)
+		icspList, err = s.addICSPsFromFile()
+		if err != nil {
+			return err
+		}
+	} else {
+		// log errors from accessing cluster since this function has no way of knowing whether it should or should not succeed
+		klog.V(5).Infof("Reading ICSP from cluster")
+		icspList, err = s.addICSPsFromCluster(ctx)
+		if err != nil {
+			klog.V(4).Infof("No alternative image sources gathered from cluster: %v", err)
+		}
+	}
+	imageRefList, err := s.alternativeImageSources(imageRef, icspList)
+	if err != nil {
+		return err
+	}
+	s.alternates = imageRefList
+	return nil
+}
+
+// alternativeImageSources returns unique list of DockerImageReference objects from list of ImageContentSourcePolicy objects
+func (s *simpleLookupICSP) alternativeImageSources(imageRef reference.DockerImageReference, icspList []operatorv1alpha1.ImageContentSourcePolicy) ([]reference.DockerImageReference, error) {
+	var imageSources []reference.DockerImageReference
+	// make sure at least 1 imagesource
+	// ie, make sure the image passed is included in image sources
+	// this is so the user-given image ref will be tried
+	imageSources = append(imageSources, imageRef.AsRepository())
+	klog.V(5).Infof("%v ImageReference added to potential ImageSourcePrefixes from ImageContentSourcePolicy", imageRef.AsRepository())
+	for _, icsp := range icspList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		for _, rdm := range repoDigestMirrors {
+			var err error
+			rdmSourceRef, err := reference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() != rdmSourceRef.AsRepository() {
```

Comment:
The way ICSP actually works when pulling, configuring {Source: `quay.io/foo`, Mirrors: `example.local/mirror-of-quay-foo`} will also match `quay.io/foo/bar` and rewrite to `example.local/mirror-of-quay-foo/bar`.

Reply:
I've added a TODO there, we'll tackle this in the next revisions.